### PR TITLE
dev/reference/sql: add document for table partition

### PR DESCRIPTION
**Pull request by tiancaiamao**
Add document for table partition

### CODE REVIEW

File: dev/reference/sql/partitioning.md [hunk 1]
```diff
@@ -0,0 +1,698 @@
+---
```
Comment (lilin90) [inline dev/reference/sql/partitioning.md]: Please add it to `TOC.md`.

File: dev/reference/sql/partitioning.md [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+title: 分区表
+category: reference
+aliases: ['/docs-cn/sql/partitioning/']
```
Comment (lilin90) [inline dev/reference/sql/partitioning.md]: If this is a newly added file, please remove the `aliases` line.

File: dev/reference/sql/partitioning.md [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+---
+
+# 分区类型
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]: Please use a sentence or two to introduce what we're going to cover in this document. This helps users decide whether they should read it.

File: dev/reference/sql/partitioning.md [hunk 4]
```diff
@@ -0,0 +1,698 @@
...
+```
+
+很显然，结果必然是在分区 `p0` 或者 `p3` 里面，也就是说，我们只需要在 `p0` 和 `p3` 里面去搜索匹配的行。去掉不必要的分区就是所谓的裁剪。优化器如果能裁剪掉一部分的分区，则执行会快于相同的查询去处理整个不做分区的表。
+
+优化器可以通过 where 条件裁剪的两个场景：
+
+* partition_column = constant
+* partition_column IN (constant1, constant2, ..., constantN)
+
+
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
```

File: dev/reference/sql/partitioning.md [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+# 分区选择
+
+SELECT 语句中支持分区选择。实现通过使用一个 `PARTITION` 选项。
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
SELECT 语句中支持分区选择。分区选择通过使用 `PARTITION` 选项实现。
```

File: dev/reference/sql/partitioning.md [hunk 6]
```diff
@@ -0,0 +1,698 @@
...
+    ('', 'Mary', 'Jones', 1, 1), ('', 'Linda', 'Black', 2, 3),
+    ('', 'Ed', 'Jones', 2, 1), ('', 'June', 'Wilson', 3, 1),
+    ('', 'Andy', 'Smith', 1, 3), ('', 'Lou', 'Waters', 2, 4),
+    ('', 'Jill', 'Stone', 1, 4), ('', 'Roger', 'White', 3, 2),
+    ('', 'Howard', 'Andrews', 1, 2), ('', 'Fred', 'Goldberg', 3, 3),
+    ('', 'Barbara', 'Brown', 2, 3), ('', 'Alice', 'Rogers', 2, 2),
+    ('', 'Mark', 'Morgan', 3, 3), ('', 'Karen', 'Cole', 3, 2);
+```
+
+
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
```

File: dev/reference/sql/partitioning.md [hunk 7]
```diff
@@ -0,0 +1,698 @@
...
+|  7 | Ed    | Jones  |        2 |             1 |
+|  8 | June  | Wilson |        3 |             1 |
+|  9 | Andy  | Smith  |        1 |             3 |
++----|-------|--------|----------|---------------+
+5 rows in set (0.00 sec)
+```
+
+如果希望获得多个分区中的行，可以提供分区名的列表，用逗号隔开。例如，`SELECT * FROM employees PARTITION (p1, p2)` 返回分区 `p1` 和 `p2` 的所有行。
+
+使用分区选择时，仍然可以使用 where 条件，ORDER BY 和 LIMIT 等选项。并且使用 HAVING 和 GROUP BY 等聚合选项也是可以的。
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
使用分区选择时，仍然可以使用 where 条件，以及 ORDER BY 和 LIMIT 等选项。使用 HAVING 和 GROUP BY 等聚合选项也是支持的。
```

File: dev/reference/sql/partitioning.md [hunk 8]
```diff
@@ -0,0 +1,698 @@
...
++---|----------+
+| c | store_id |
++---|----------+
+| 5 |        2 |
+| 5 |        3 |
++---|----------+
+2 rows in set (0.00 sec)
+```
+
+分支选择支持所有类型的分区表，无论是 range 分区或是 hash 分区等。对于 hash 分区，如果没有指定分区名，会自动使用 `p0,p1,p2...pN-1** 作为分区名。
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
分支选择支持所有类型的分区表，无论是 range 分区或是 hash 分区等。对于 hash 分区，如果没有指定分区名，会自动使用 `p0`、`p1`、`p2`……或 `pN-1` 作为分区名。
```

File: dev/reference/sql/partitioning.md [hunk 9]
```diff
@@ -0,0 +1,698 @@
...
+
+在 `INSERT ... SELECT** 的 `SELECT** 中也是可以使用分区选择的。
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
在 `INSERT ... SELECT` 的 `SELECT` 中也是支持分区选择的。
```

File: dev/reference/sql/partitioning.md [hunk 10]
```diff
@@ -0,0 +1,698 @@
...
+
+# 分区的约束和限制
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
# 分区的约束和限制
```
Please make it a second-level heading.

File: dev/reference/sql/partitioning.md [hunk 11]
```diff
@@ -0,0 +1,698 @@
...
+
+这一节介绍当前 TiDB 分区表的一些约束和限制。
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
本节介绍当前 TiDB 分区表的一些约束和限制。
```

File: dev/reference/sql/partitioning.md [hunk 12]
```diff
@@ -0,0 +1,698 @@
...
+
+## 分区键，主键和唯一键
+
+这一节讨论分区键，主键和唯一键之间的关系。一句话总结它们之间的关系要满足的规则：**分区表的每个唯一列，必须包含分区表达式中用到的所有列**
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
本节讨论分区键、主键和唯一键之间的关系。一句话总结它们之间的关系要满足的规则：**分区表的每个唯一列，必须包含分区表达式中用到的所有列**。
```

File: dev/reference/sql/partitioning.md [hunk 13]
```diff
@@ -0,0 +1,698 @@
...
+
+>  every unique key on the table must use every column in the table's partitioning expression
+
+这里所指的唯一也包含了主键，因为根据主键的定义，主键必须是唯一的。举个例子，下面这些建表语句就是无效的：
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
这里所指的唯一也包含了主键，因为根据主键的定义，主键必须是唯一的。例如，下面这些建表语句就是无效的：
```

File: dev/reference/sql/partitioning.md [hunk 14]
```diff
@@ -0,0 +1,698 @@
...
+    col3 INT NOT NULL,
+    col4 INT NOT NULL,
+    UNIQUE KEY (col1),
+    UNIQUE KEY (col3)
+)
+PARTITION BY HASH(col1 + col3)
+PARTITIONS 4;
+```
+
+它们都是有唯一键没有包含所有分区键的。
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
它们都是有唯一键但没有包含所有分区键的。
```

File: dev/reference/sql/partitioning.md [hunk 15]
```diff
@@ -0,0 +1,698 @@
...
+    col2 DATE NOT NULL,
+    col3 INT NOT NULL,
+    col4 INT NOT NULL,
+    UNIQUE KEY (col1, col3)
+)
+PARTITION BY HASH(col1 + col3)
+PARTITIONS 4;
+```
+
+这个例子会产生一个报错：
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
上例中会产生一个报错：
```

File: dev/reference/sql/partitioning.md [hunk 16]
```diff
@@ -0,0 +1,698 @@
...
+
+```
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
```sql
```

File: dev/reference/sql/partitioning.md [hunk 17]
```diff
@@ -0,0 +1,698 @@
...
+    PRIMARY KEY(col1, col3),
+    UNIQUE KEY(col2)
+)
+PARTITION BY HASH( YEAR(col2) )
+PARTITIONS 4;
+```
+
+两个例子中，主键都没有包含分区表达式中的全部的列。
+
+如果既没有主键，也没有唯一键，则这个限制就没有了。
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
如果既没有主键，也没有唯一键，则不存在这个限制。
```

File: dev/reference/sql/partitioning.md [hunk 18]
```diff
@@ -0,0 +1,698 @@
...
+  partition p1 values less than (7),
+  partition p2 values less than (11));
+Query OK, 0 rows affected (0.10 sec)
+
+mysql> insert into t values (1, 2), (3, 4),(5, 6),(7,8),(9,10);
+Query OK, 5 rows affected (0.01 sec)
+Records: 5  Duplicates: 0  Warnings: 0
+```
+
+TiDB 每次返回结果会不同
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]:
```suggestion
TiDB 每次返回结果会不同，例如：
```

File: dev/reference/sql/partitioning.md [hunk 19]
```diff
@@ -0,0 +1,698 @@
...
+TIME_TO_SEC()
+TO_DAYS()
+TO_SECONDS()
+UNIX_TIMESTAMP() (with TIMESTAMP columns)
+WEEKDAY()
+YEAR()
+YEARWEEK()
+```
+
+## 兼容性
```
Comment (yikeke) [inline dev/reference/sql/partitioning.md]: 考虑用 unordered list 列出这节的兼容性要点，更方便读者查看。

File: dev/reference/sql/partitioning.md [hunk 20]
```diff
@@ -0,0 +1,698 @@
...
+
+* 删除旧数据。如果你使用之前的 `employees` 表的例子，你可以简单使用 `ALTER TABLE employees DROP PARTITION p0;` 删除所有在 1991 年以前停止继续在这家公司工作的员工记录。这会比使用 `DELETE FROM employees WHERE YEAR(separated) <= 1990;` 执行快得多。
+* 使用包含时间或者日期的列，或者是其它按序生成的数据。
+* 频繁查询分区使用的列。例如执行这样的查询 `EXPLAIN SELECT COUNT(*) FROM employees WHERE separated BETWEEN '2000-01-01' AND '2000-12-31' GROUP BY store_id;`，TiDB 可以迅速确定，只需要扫描 `p2` 分区的数据，因为其它的分区不满足 `where` 条件。
+
+## Hash 分区
+
+Hash 分区主要用于保证数据均匀地分散到一定数量的分区里面。在 range 分区中你必须为每个分区指定值的范围；在 hash 分区中，你只需要指定分区的数量。
+
+使用 hash 分区时，需要在 `CREATE TABLE` 后面添加 `PARTITION BY HASH (expr)`，其中 `expr` 是一个返回整数的表达式。它可以是一个列名，如果这一列的类型是整数类型。此外，你大概还需要加上 `PARTITIONS num`，其中 `num` 是一个正整数，表示将表划分多少分区。
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply dev/reference/sql/partitioning.md]: 这里英语单词是 "probably"，我把中文改成“很可能”了

File: dev/reference/sql/partitioning.md [hunk 21]
```diff
@@ -0,0 +1,698 @@
...
+)
+PARTITION BY HASH( YEAR(hired) )
+PARTITIONS 4;
+```
+
+最高效的 hash 函数是作用在单列上，并且函数的单调性是跟列的值一样递增或者递减的，因为这种情况可以像 range 分区一样裁剪。
+
+例如 `date_col` 是类型为 `DATE` 的列，表达式 `TO_DAYS(date_col)` 的值是直接随 `date_col` 的值变化的。`YEAR(date_col)` 跟 `TO_DAYS(date_col)` 就不太一样，因为不是每次 `date_col` 变化时 `YEAR(date_col)` 都会得到不同的值。即使如此，`YEAR(date_col)` 也仍然是一个比如好的 hash 函数，因为它的结果是随着 `date_col` 的值的比例变化的。
+
+作为对比，假设我们有一个类型是 INT 的 `int_col` 的列。考虑一下这个表达式 `POW(5-int_col,3) + 6`，这并不是一个比较好的 hash 函数，因为 `int_col` 的值的变化，变化会让表达式的结果成比例地变化。改变 `int_col` 的值会使表达式的结果的值变化巨大。例如，`int_col` 从 5 变到 6 表达式的结果变化是 -1，但是从 6 变到 7 的时候表达式的值的变化是 -7。
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply dev/reference/sql/partitioning.md]: 对，这里写错了

File: dev/reference/sql/partitioning.md (hunk 15 above)
Comment (tiancaiamao) [reply dev/reference/sql/partitioning.md]: 是下例呀

File: dev/reference/sql/partitioning.md (hunk 1 above)
Comment (tiancaiamao) [reply dev/reference/sql/partitioning.md]: Done

File: dev/reference/sql/partitioning.md [hunk 22]
```diff
@@ -0,0 +1,698 @@
...
+
+总而言之，表达式越接近 `y = cx` 的形式，它越是适合作为 hash 函数。因为表达式越是非线性的，在各个分区上面的数据的分布越是倾向于不均匀。
+
+理论上，hash 分区也是可以做分区裁剪的。而实际上对于多列的情况，实现很难并且计算很耗时。因此，不推荐 hash 分区在表达式中涉及多列。
+
+使用 `PARTITIION BY HASH` 的时候，TiDB 通过表达式的结果做“取余”运算，决定数据落在哪个分区。换句话说，如果分区表达式是 `expr`，分区数是 `num`，则由 `MOD(expr, num)` 决定存储的分区。假设 `t1` 定义如下：
+
+```
+`CREATE TABLE t1 (col1 INT, col2 CHAR(5), col3 DATE)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply dev/reference/sql/partitioning.md]: 不需要

**Pull request merged**